crates/re_log_types/src/data_cell.rs

```diff
@@ -0,0 +1,445 @@
+/// Cloning a `DataCell` is thus cheap (shallow, ref-counted).
+///
+/// ## Layout
+///
+/// A cell is an array of component instances: `[C, C, C, ...]`.
+///
+/// Consider this example:
+/// ```ignore
+/// let points: &[Point2D] = &[
+///     [10.0, 10.0].into(),
+///     [20.0, 20.0].into(),
+///     [30.0, 30.0].into(),
+/// ];
+/// let cell = DataCell::from(points);
+/// // Or, alternatively:
+/// let cell = DataCell::from_component::<Point2D>([[10.0, 10.0], [20.0, 20.0], [30.0, 30.0]]);
+/// ```
+///
+/// The cell's datatype is now a `StructArray`:
+/// ```ignore
+/// Struct([
+///    Field { name: "x", data_type: Float32, is_nullable: false, metadata: {} },
+///    Field { name: "y", data_type: Float32, is_nullable: false, metadata: {} },
+/// ])
+/// ```
+///
+/// Or, visualized as a cell within a larger table:
+/// ```text
+/// ┌──────────────────────────────────────────────────┐
+/// │ rerun.point2d                                    │
+/// ╞══════════════════════════════════════════════════╡
+/// │ [{x: 10, y: 10}, {x: 20, y: 20}, {x: 30, y: 30}] │
+/// └──────────────────────────────────────────────────┘
+/// ```
+///
+/// ## Example
+///
+/// ```rust
+/// # use arrow2_convert::field::ArrowField as _;
+/// # use itertools::Itertools as _;
+/// #
+/// # use re_log_types::{DataCell, msg_bundle::Component as _};
+/// # use re_log_types::component_types::Point2D;
+/// #
+/// let points: &[Point2D] = &[
+///     [10.0, 10.0].into(),
+///     [20.0, 20.0].into(),
+///     [30.0, 30.0].into(),
+/// ];
+/// let _cell = DataCell::from(points);
+///
+/// // Or, alternatively:
+/// let cell = DataCell::from_component::<Point2D>([[10.0, 10.0], [20.0, 20.0], [30.0, 30.0]]);
+///
+/// eprintln!("{:#?}", cell.datatype());
+/// eprintln!("{cell}");
+/// #
+/// # assert_eq!(Point2D::name(), cell.component());
+/// # assert_eq!(3, cell.num_instances());
+/// # assert_eq!(cell.datatype(), &Point2D::data_type());
+/// #
+/// # assert_eq!(points, cell.as_native().collect_vec().as_slice());
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct DataCell {
+    /// Name of the component type used in this cell.
+    //
+    // TODO(cmc): We need to store this extra information within the values array itself,
+    // rather than outside of it. Arrow has the concept of extensions specifically for storing
+    // type metadata, but we have had some issues with it in the past. This is an opportunity to
+    // revisit and improve upon that implementation.
+    pub(crate) name: ComponentName,
+
+    /// A uniformly typed list of values for the given component type: `[C, C, C, ...]`
+    ///
+    /// Includes the data, its schema and probably soon the component metadata
+    /// (e.g. the `ComponentName`).
+    ///
+    /// Internally this is always stored as an erased arrow array to avoid bad surprises with
+    /// frequent boxing/unboxing down the line.
+    pub(crate) values: Box<dyn arrow2::array::Array>,
```

Comment:
I think it worth noting here that this `Array` can be a slice of another `Array`